### PR TITLE
python: fix usability of curses.h

### PR DESCRIPTION
**Pull request by XuehaiPan**
- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/linuxbrew-core/blob/HEAD/CONTRIBUTING.md)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/linuxbrew-core/pulls) for the same formula update/change?
- [X] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

The headers of `curses` can be found but cannot be compiled. _(It's really weird that all headers in `#{Formula["ncurses"].opt_include}` are already symlinked to `#{HOMEBREW_PREFIX}/include`.)_

Partial output of `brew install --verbose --debug python3`:

```
checking curses.h usability... no
checking curses.h presence... yes
configure: WARNING: curses.h: present but cannot be compiled
configure: WARNING: curses.h:     check for missing prerequisite headers?
configure: WARNING: curses.h: see the Autoconf documentation
configure: WARNING: curses.h:     section "Present But Cannot Be Compiled"
configure: WARNING: curses.h: proceeding with the compiler's result
configure: WARNING:     ## --------------------------------------- ##
configure: WARNING:     ## Report this to https://bugs.python.org/ ##
configure: WARNING:     ## --------------------------------------- ##
checking for curses.h... no
checking ncurses.h usability... no
checking ncurses.h presence... yes
configure: WARNING: ncurses.h: present but cannot be compiled
configure: WARNING: ncurses.h:     check for missing prerequisite headers?
configure: WARNING: ncurses.h: see the Autoconf documentation
configure: WARNING: ncurses.h:     section "Present But Cannot Be Compiled"
configure: WARNING: ncurses.h: proceeding with the compiler's result
configure: WARNING:     ## --------------------------------------- ##
configure: WARNING:     ## Report this to https://bugs.python.org/ ##
configure: WARNING:     ## --------------------------------------- ##
checking for ncurses.h... no
checking for term.h... yes
```

I add `-I#{Formula["ncurses"].opt_include}` to `cppflags` explicitly. Partial output of `brew install --verbose --debug python3` after adding new entry:

```
checking curses.h usability... yes
checking curses.h presence... yes
checking for curses.h... yes
checking ncurses.h usability... yes
checking ncurses.h presence... yes
checking for ncurses.h... yes
checking for term.h... yes
checking whether mvwdelch is an expression... yes
checking whether WINDOW has _flags... yes
checking for is_pad... yes
checking for is_term_resized... yes
checking for resize_term... yes
checking for resizeterm... yes
checking for immedok... yes
checking for syncok... yes
checking for wchgat... yes
checking for filter... yes
checking for has_key... yes
checking for typeahead... yes
checking for use_env... yes
```

### CODE REVIEW

**Comment by XuehaiPan** [comment]
Screenshots of master version (`CPPFLAGS="-I. -I/Include -I${HOMEBREW_PREFIX}/include"`):

Screenshot config master
Screenshot python master

**Comment by XuehaiPan** [comment]
Screenshots of master version (`CPPFLAGS="-I. -I/Include -I${HOMEBREW_PREFIX}/include -I$(brew --prefix ncurses)/include"`):

Screenshot config fix
Screenshot python fix

**Comment by XuehaiPan** [comment]
This might be a bug in `configure.ac` of `cpython`.
> 
> The headers of `curses` can be found but cannot be compiled. _(It's really weird that all headers in `#{Formula["ncurses"].opt_include}` are already symlinked to `#{HOMEBREW_PREFIX}/include`.)_
> 

Screenshot

**Comment by XuehaiPan** [comment]
`configure` can found and compile `curses.h` successfully after unsetting `C_INCLUDE_PATH` in official `cpython` repo.

Screen Shot

But the issue still exists after I removing the environment variables in `Formula/python@3.9.rb`. I added the following lines to `install` section in `Formula/python@3.9.rb`:

```
ENV.delete("CPATH")
ENV.delete("C_INCLUDE_PATH")
ENV.delete("CPLUS_INCLUDE_PATH")
```

**Comment by danielnachun** [comment]
Thanks for trying to fix you issue with a pull request.  Please avoid using screenshots as they are much harder to read than text pasted in the screen.

Before we go too far down a rabbit hole here, can you try to install `python@3.9` again with no modifications to the formula or Python itself?  And after the install fails, run `brew gist-logs python@3.9` and paste the link that it generates here.  I'd like to look at `config.log` generated by that to see what's actually driving the `present but cannot be compiled error`.  In many cases these errors can be misleading, and are actually driven by unrelated problems.

`python3` is a very popular package which is probably installed by almost every Homebrew user on Linux, so I suspect there is something about your particular setup which is causing this or we would be seeing this error more often.   I personally use Homebrew on 3 different HPC systems, and have never had any issues with getting Python installed.

**Comment by XuehaiPan** [comment]
`brew install python3` generates no errors and passed all test with no modifications to the formula. But it raises `no-member` errors when I using `curses` module with that brew-installed python3.

I removed `~/.linuxbrew` and made a brand new Linuxbrew installation. Partial output of `brew install --verbose python3`:

```
$ brew install --verbose python3
##### ...
##### installation of dependencies #####
##### ...
==> Installing python@3.9
tar xof /home/panxuehai/.cache/Homebrew/downloads/4f82a87872361431c1ff89331abe8b4474053a6310159862948c4e0f35bdaca5--Python-3.9.1.tar.xz -C /tmp/d20210218-59859-g8rnzm
cp -pR /tmp/d20210218-59859-g8rnzm/Python-3.9.1/. /tmp/python-3.9-20210218-59859-12tw06q/Python-3.9.1
chmod -Rf +w /tmp/d20210218-59859-g8rnzm
==> ./configure --prefix=/home/panxuehai/.linuxbrew/Cellar/python@3.9/3.9.1_9 --enable-ipv6 --datarootdir=/home/panxuehai/.linuxbrew/Cellar/python@3.9/3.9.1_9/share --datadir=/home/panxuehai/.linuxbrew/Cellar/python@3.9/3.9.1_9/share --enable-loadable-sqlite-extensions --with-openssl=/home/panxuehai/.linuxbrew/opt/openssl@1.1 --with-dbmliborder=gdbm:ndbm --enable-optimizations --with-lto --enable-shared --with-system-ffi MACOSX_DEPLOYMENT_TARGET= --with-tcltk-includes=-I/home/panxuehai/.linuxbrew/opt/tcl-tk/include --with-tcltk-libs=-L/home/panxuehai/.linuxbrew/opt/tcl-tk/lib -ltcl8.6 -ltk8.6 CFLAGS_NODIST=-I/home/panxuehai/.linuxbrew/include LDFLAGS_NODIST=-L/home/panxuehai/.linuxbrew/lib CPPFLAGS=-I/home/panxuehai/.linuxbrew/include
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking for python3.9... no
checking for python3... python3
##### ...
##### many lines of configure output #####
##### ...
checking for tv_nsec2 in struct stat... no
checking curses.h usability... no
checking curses.h presence... yes
configure: WARNING: curses.h: present but cannot be compiled
configure: WARNING: curses.h:     check for missing prerequisite headers?
configure: WARNING: curses.h: see the Autoconf documentation
configure: WARNING: curses.h:     section "Present But Cannot Be Compiled"
configure: WARNING: curses.h: proceeding with the compiler's result
configure: WARNING:     ## --------------------------------------- ##
configure: WARNING:     ## Report this to https://bugs.python.org/ ##
configure: WARNING:     ## --------------------------------------- ##
checking for curses.h... no
checking ncurses.h usability... no
checking ncurses.h presence... yes
configure: WARNING: ncurses.h: present but cannot be compiled
configure: WARNING: ncurses.h:     check for missing prerequisite headers?
configure: WARNING: ncurses.h: see the Autoconf documentation
configure: WARNING: ncurses.h:     section "Present But Cannot Be Compiled"
configure: WARNING: ncurses.h: proceeding with the compiler's result
configure: WARNING:     ## --------------------------------------- ##
configure: WARNING:     ## Report this to https://bugs.python.org/ ##
configure: WARNING:     ## --------------------------------------- ##
checking for ncurses.h... no
checking for term.h... yes
checking whether mvwdelch is an expression... no
checking whether WINDOW has _flags... no
checking for is_pad... no
checking for is_term_resized... no
checking for resize_term... no
checking for resizeterm... no
checking for immedok... no
checking for syncok... no
checking for wchgat... no
checking for filter... no
checking for has_key... no
checking for typeahead... no
checking for use_env... no
configure: checking for device files
##### ...
##### the remaining lines of brew-install output #####
##### ...
==> python@3.9
Python has been installed as
  /home/panxuehai/.linuxbrew/bin/python3

Unversioned symlinks `python`, `python-config`, `pip` etc. pointing to
`python3`, `python3-config`, `pip3` etc., respectively, have been installed into
  /home/panxuehai/.linuxbrew/opt/python@3.9/libexec/bin

You can install Python packages with
  pip3 install <package>
They will install into the site-package directory
  /home/panxuehai/.linuxbrew/lib/python3.9/site-packages

See: https://docs.brew.sh/Homebrew-and-Python
```

`brew install --verbose python3` exited with 0.

My OS is `Ubuntu 16.04 LTS`, and the environment variables:

```
CC=/usr/bin/gcc   # gcc-6.5
CXX=/usr/bin/g++   # g++-6.5
C_INCLUDE_PATH=/home/panxuehai/.linuxbrew/include
CPLUS_INCLUDE_PATH=/home/panxuehai/.linuxbrew/include
```

**Comment by XuehaiPan** [comment]
Results of `brew gist-log python3` in a newly installed Linuxbrew: https://gist.github.com/XuehaiPan/946ff6ee675dd9552a969ec03c9c2831

**Comment by XuehaiPan** [comment]
Output of `"$(brew --prefix python@3.9)/bin/python3" -c "import sysconfig; [print(key, repr(value)) for key, value in sysconfig.get_config_vars().items()]" | grep -i -E '(^\w*(curses|((config|c|ld)\w*(args|flags)))\w*)|curses'`:

```
BASECFLAGS '-Wno-unused-result -Wsign-compare'
BASECPPFLAGS ''
CFLAGS '-Wno-unused-result -Wsign-compare -DNDEBUG -g  -O3 -Wall'
CFLAGSFORSHARED '-fPIC'
CFLAGS_ALIASING ''
CONFIGURE_CFLAGS ''
CONFIGURE_CFLAGS_NODIST '-I/home/panxuehai/.linuxbrew/include -flto -fuse-linker-plugin -ffat-lto-objects -flto-partition=none -g -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Werror=implicit-function-declaration -fvisibility=hidden'
CONFIGURE_CPPFLAGS '-I/home/panxuehai/.linuxbrew/include'
CONFIGURE_LDFLAGS ''
CONFIGURE_LDFLAGS_NODIST '-L/home/panxuehai/.linuxbrew/lib -flto -fuse-linker-plugin -ffat-lto-objects -flto-partition=none -g'
CONFIG_ARGS "'--prefix=/home/panxuehai/.linuxbrew/Cellar/python@3.9/3.9.1_9' '--enable-ipv6' '--datarootdir=/home/panxuehai/.linuxbrew/Cellar/python@3.9/3.9.1_9/share' '--datadir=/home/panxuehai/.linuxbrew/Cellar/python@3.9/3.9.1_9/share' '--enable-loadable-sqlite-extensions' '--with-openssl=/home/panxuehai/.linuxbrew/opt/openssl@1.1' '--with-dbmliborder=gdbm:ndbm' '--enable-optimizations' '--with-lto' '--enable-shared' '--with-system-ffi' 'MACOSX_DEPLOYMENT_TARGET=' '--with-tcltk-includes=-I/home/panxuehai/.linuxbrew/opt/tcl-tk/include' '--with-tcltk-libs=-L/home/panxuehai/.linuxbrew/opt/tcl-tk/lib -ltcl8.6 -ltk8.6' 'CFLAGS_NODIST=-I/home/panxuehai/.linuxbrew/include' 'LDFLAGS_NODIST=-L/home/panxuehai/.linuxbrew/lib' 'CPPFLAGS=-I/home/panxuehai/.linuxbrew/include' 'CC=gcc-5' 'PKG_CONFIG_PATH=/home/panxuehai/.linuxbrew/opt/openssl@1.1/lib/pkgconfig:/home/panxuehai/.linuxbrew/opt/ncurses/lib/pkgconfig:/home/panxuehai/.linuxbrew/opt/readline/lib/pkgconfig:/home/panxuehai/.linuxbrew/opt/zlib/lib/pkgconfig:/home/panxuehai/.linuxbrew/opt/sqlite/lib/pkgconfig:/home/panxuehai/.linuxbrew/opt/libpthread-stubs/lib/pkgconfig:/home/panxuehai/.linuxbrew/opt/libxau/lib/pkgconfig:/home/panxuehai/.linuxbrew/opt/libxdmcp/lib/pkgconfig:/home/panxuehai/.linuxbrew/opt/libxcb/lib/pkgconfig:/home/panxuehai/.linuxbrew/opt/libx11/lib/pkgconfig:/home/panxuehai/.linuxbrew/opt/libxext/lib/pkgconfig:/home/panxuehai/.linuxbrew/opt/tcl-tk/lib/pkgconfig:/home/panxuehai/.linuxbrew/opt/xz/lib/pkgconfig:/home/panxuehai/.linuxbrew/opt/libffi/lib/pkgconfig:/home/panxuehai/.linuxbrew/opt/xorgproto/share/pkgconfig'"
CPPFLAGS '-I. -I./Include -I/home/panxuehai/.linuxbrew/include'
EXTRA_CFLAGS ''
HAVE_CHFLAGS 0
HAVE_CURSES_FILTER 0
HAVE_CURSES_H 0
HAVE_CURSES_HAS_KEY 0
HAVE_CURSES_IMMEDOK 0
HAVE_CURSES_IS_PAD 0
HAVE_CURSES_IS_TERM_RESIZED 0
HAVE_CURSES_RESIZETERM 0
HAVE_CURSES_RESIZE_TERM 0
HAVE_CURSES_SYNCOK 0
HAVE_CURSES_TYPEAHEAD 0
HAVE_CURSES_USE_ENV 0
HAVE_CURSES_WCHGAT 0
HAVE_LCHFLAGS 0
HAVE_NCURSES_H 0
HAVE_STRUCT_STAT_ST_FLAGS 0
LDFLAGS ''
MULTIARCH_CPPFLAGS '-DMULTIARCH=\\"x86_64-linux-gnu\\"'
OPENSSL_LDFLAGS '-L/home/panxuehai/.linuxbrew/opt/openssl@1.1/lib'
PY_BUILTIN_MODULE_CFLAGS '-Wno-unused-result -Wsign-compare -DNDEBUG -g  -O3 -Wall -I/home/panxuehai/.linuxbrew/include -flto -fuse-linker-plugin -ffat-lto-objects -flto-partition=none -g -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Werror=implicit-function-declaration -fvisibility=hidden -fprofile-use -fprofile-correction -I./Include/internal -I. -I./Include -I/home/panxuehai/.linuxbrew/include -fPIC -DPy_BUILD_CORE_BUILTIN'
PY_CFLAGS '-Wno-unused-result -Wsign-compare -DNDEBUG -g  -O3 -Wall'
PY_CFLAGS_NODIST '-I/home/panxuehai/.linuxbrew/include -flto -fuse-linker-plugin -ffat-lto-objects -flto-partition=none -g -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Werror=implicit-function-declaration -fvisibility=hidden -fprofile-use -fprofile-correction -I./Include/internal'
PY_CORE_CFLAGS '-Wno-unused-result -Wsign-compare -DNDEBUG -g  -O3 -Wall -I/home/panxuehai/.linuxbrew/include -flto -fuse-linker-plugin -ffat-lto-objects -flto-partition=none -g -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Werror=implicit-function-declaration -fvisibility=hidden -fprofile-use -fprofile-correction -I./Include/internal -I. -I./Include -I/home/panxuehai/.linuxbrew/include -fPIC -DPy_BUILD_CORE'
PY_CORE_LDFLAGS '-L/home/panxuehai/.linuxbrew/lib -flto -fuse-linker-plugin -ffat-lto-objects -flto-partition=none -g'
PY_CPPFLAGS '-I. -I./Include -I/home/panxuehai/.linuxbrew/include'
PY_LDFLAGS ''
PY_LDFLAGS_NODIST '-L/home/panxuehai/.linuxbrew/lib -flto -fuse-linker-plugin -ffat-lto-objects -flto-partition=none -g'
PY_STDMODULE_CFLAGS '-Wno-unused-result -Wsign-compare -DNDEBUG -g  -O3 -Wall -I/home/panxuehai/.linuxbrew/include -flto -fuse-linker-plugin -ffat-lto-objects -flto-partition=none -g -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Werror=implicit-function-declaration -fvisibility=hidden -fprofile-use -fprofile-correction -I./Include/internal -I. -I./Include -I/home/panxuehai/.linuxbrew/include -fPIC'
STRICT_SYSV_CURSES "/* Don't use ncurses extensions */"
```

**Comment by danielnachun** [comment]
This is why the you're getting the `present but cannot be compiled` errors:
```
configure:16155: gcc-5 -pthread -c  -I/home/panxuehai/.linuxbrew/include -I/usr/include/ncursesw conftest.c >&5
In file included from conftest.c:405:0:
/home/panxuehai/.linuxbrew/include/ncursesw/unctrl.h:61:38: error: conflicting types for 'unctrl'
 NCURSES_EXPORT(NCURSES_CONST char *) NCURSES_SP_NAME(unctrl) (SCREEN*, chtype);
                                      ^
In file included from /usr/include/ncursesw/ncurses.h:2012:0,
                 from conftest.c:405:
/home/panxuehai/.linuxbrew/include/ncursesw/unctrl.h:58:38: note: previous declaration of 'unctrl' was here
 NCURSES_EXPORT(NCURSES_CONST char *) unctrl (chtype);
```

For some reason you still have `-I/usr/include/ncursesw` in your CFLAGS, and this is causing your brewed `ncurses` to try to use header files from the system `ncurses`.  This causes a failure because the brewed `ncurses` is a different version from the system one, and the build will fail if you mix headers from the two.  System folders like `-I/usr/include` aren't supposed to show up in the `CFLAGS` when homebrew is running.  

Can you post the output of `pkg-config --cflags ncurses`?  We'll see if something weird is going on with your brewed `ncurses`.

**Comment by XuehaiPan** [comment]
```
$ which -a pkg-config
/home/panxuehai/.linuxbrew/bin/pkg-config
/usr/bin/pkg-config

$ "$(brew --prefix)/bin/pkg-config" --cflags ncurses                                                                                                                                                                         
-D_DEFAULT_SOURCE -D_XOPEN_SOURCE=600 -I/home/panxuehai/.linuxbrew/Cellar/ncurses/6.2/include/ncursesw -I/home/panxuehai/.linuxbrew/Cellar/ncurses/6.2/include

$ /usr/bin/pkg-config --cflags ncurses
-D_GNU_SOURCE
```

**Comment by XuehaiPan** [comment]
### Reproduce this issue step-by-step on `ubuntu:16.04` using `docker`.

**1. Create a file named `Dockerfile` with the following content:**
```dockerfile
FROM ubuntu:16.04

# Setup Ubuntu
ENV DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -yq ca-certificates
RUN apt-get update && apt-get install -yq apt-utils
RUN apt-get install -yq sudo build-essential curl file git

# Install ncurses5
RUN apt-get install -yq libncurses5-dev libncursesw5-dev lib32ncursesw5-dev

# Add non-root user
RUN useradd --create-home --password brew --shell /bin/bash brew-test
USER brew-test
RUN cd /home/brew-test

# Install Linuxbrew
RUN /bin/bash -c "$( \
    curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh | \
    sed -E 's|^(\s*HOMEBREW_PREFIX_DEFAULT)=.*$|\1="/home/brew-test/.linuxbrew"|g' \
  )"
ENV PATH="/home/brew-test/.linuxbrew/bin:${PATH}"
ENV HOMEBREW_PREFIX="/home/brew-test/.linuxbrew"
ENV HOMEBREW_CELLAR="/home/brew-test/.linuxbrew/Cellar"
RUN echo 'eval "$(/home/brew-test/.linuxbrew/bin/brew shellenv)"' >> /home/brew-test/.profile

# Install dependencies of python3
RUN brew update --quiet
RUN brew install --quiet gcc $(brew deps python3)

# Do cleanup
USER root
RUN apt-get autoclean
USER brew-test
RUN brew cleanup --quiet -s --prune 0

WORKDIR '/home/brew-test'
CMD [ "/bin/bash", "--login" ]
```

**2. Build docker image (may take several minutes):**
```
# On host
user@host:~$ docker build -f Dockerfile . -t brew-python3
```

**3. Install the `python3` formula using Linuxbrew:**
```
# On host
user@host:~$ docker run -it -h ubuntu brew-python3

# In docker container
brew-test@ubuntu:~$ brew install --verbose python3
```

**Comment by XuehaiPan** [comment]
I found out that `configure` from `cpython` always add `-I/usr/include/ncursesw` to `CPPFLAGS` when doing curses header check.

[`cpython/blob/3.9/configure#L16141-L16581`](https://github.com/python/cpython/blob/3.9/configure#L16141-L16581):
```configure
# first curses header check
ac_save_cppflags="$CPPFLAGS"
if test "$cross_compiling" = no; then
  CPPFLAGS="$CPPFLAGS -I/usr/include/ncursesw"
fi

for ac_header in curses.h ncurses.h
do :
  as_ac_Header=`$as_echo "ac_cv_header_$ac_header" | $as_tr_sh`
ac_fn_c_check_header_mongrel "$LINENO" "$ac_header" "$as_ac_Header" "$ac_includes_default"
if eval test \"x\$"$as_ac_Header"\" = x"yes"; then :
  cat >>confdefs.h <<_ACEOF
#define `$as_echo "HAVE_$ac_header" | $as_tr_cpp` 1
_ACEOF

fi

done

##### many lines of curses header check #####

# last curses configure check
CPPFLAGS=$ac_save_cppflags
```

I replaced system ncurses include entry with the brewed one in `Formula/python@3.*.rb`:
```ruby
# We want our ncurses! Override system ncurses includes!
inreplace "configure",
  'CPPFLAGS="$CPPFLAGS -I/usr/include/ncursesw"',
  "CPPFLAGS=\"$CPPFLAGS -I#{Formula["ncurses"].opt_include}\""
```

And it works for me.

**Comment by danielnachun** [comment]
Great job investigating this so far.  I think this problem has evaded our usual checks because it requires a specific set of circumstances to occur.  The user must have the `ncurses` headers installed to `/usr/include` (our CI Docker environment does **not** have these headers, so it doesn't have any issues), and I think the version of `ncurses` needs to be 6.  I can reproduce this error in the CI Docker if I first do `sudo apt-get install libncursesw5-dev` and then try to build Python from source.

I've only built Python from source before this on two CentOS 7 systems which have ncurses 5.9 headers (which seem to be ignored), and Ubuntu 16.04 system that did have the `ncurses` headers installed.  Ironically if you didn't install nurses in your Dockerfile, that should also "fix" the problem, but that's not really a solution if the user doesn't have admin rights.

Python hardcoding an include directory like that is not a good practice, but I can see in the logs that `-I$HOMEBREW_PREFIX/include` is added before `-I/usr/include/ncursesw`.  What's really odd here from `config.log` is that for some reason `gcc` does actually use header files from brewed `ncurses`, but then arbitrarily decides to start using system `ncurses` headers as well!  I've never seen this happen before.

I haven't found a way to fix this other than changing the configure script as you have done.  This is something we obviously try to avoid doing whenever possible, but we may have no other choice here.  I'll ping some other maintainers to take a look at this first as Python is a critical formula and any changes we make to it should be carefully checked first.

**Comment by iMichka** [comment]
I think the difference between the configure step and the actual compilation is that our gcc uses the brew gcc shim, which modifies include paths on the fly. But we do not have a configure shim, so configure might pick stuff that our gcc shim will handle differently. It's not the first time this has happened.

File: Formula/python@3.9.rb [hunk 1]
```diff
@@ -16,7 +16,7 @@ class PythonAT39 < Formula
     sha256 big_sur:       "5aaf6c9ba2ddcdf9e240ad6843abbf2a7fc8df3cf58e4a08680ddc88176fd00d"
     sha256 catalina:      "65c25cb1486cbda7ce41ce05b3e22a68095b41d50956bda1c9c6eec2dc566916"
     sha256 mojave:        "162eb3c703f64fcc7b56df755ddf9d70aafe85f4e8f8af2e273de0b7c8a31065"
-    sha256 x86_64_linux:  "b19ab9c2981df2ceaa3b5be288c9c3e20a1f3a7a5dbffaf561eb02ec15df6c9d"
+    sha256 x86_64_linux:  "0ecc8baef539d18a616c62ea74a0681d7f09531a2d032bcdbf8eef803b2aff12"
```
Comment (iMichka) [inline Formula/python@3.9.rb:19]: Please leave the sha256 lines as-is. We won't rebuild bottles for python here as this impacts only people building from source.

Comment (XuehaiPan) [reply Formula/python@3.9.rb]: CI tests failed for sha256 hash changes.

Comment (iMichka) [reply Formula/python@3.9.rb]: That’s fine, I’ll merge this even if CI is red due to that. Our Linux CI is not super clever, we are working on this.

**Review by g1234567890123 — APPROVED** [review]

**Comment by danielnachun** [comment]
Closing for now due to suspicious activity.

**Comment by XuehaiPan** [comment]
Hi, is there any update for this issue?

**Comment by iMichka** [comment]
I am fine with the fix.

Can you open the pull request in https://github.com/Homebrew/homebrew-core ?
One pull request per formula.
The inreplaces need to be put in `on_linux do / end` blocks, as these changes are linux only.

Please add the failure/warning to the commit message:
```
checking for tv_nsec2 in struct stat... no
checking curses.h usability... no
checking curses.h presence... yes
configure: WARNING: curses.h: present but cannot be compiled
configure: WARNING: curses.h:     check for missing prerequisite headers?
configure: WARNING: curses.h: see the Autoconf documentation
configure: WARNING: curses.h:     section "Present But Cannot Be Compiled"
configure: WARNING: curses.h: proceeding with the compiler's result
configure: WARNING:     ## --------------------------------------- ##
configure: WARNING:     ## Report this to https://bugs.python.org/ ##
configure: WARNING:     ## --------------------------------------- ##
checking for curses.h... no
checking ncurses.h usability... no
checking ncurses.h presence... yes
```

and explain that sometimes the curses usage in Python might get broken on Linux.

**Comment by XuehaiPan** [comment]
I opened 3 PRs on [Homebrew/homebrew-core](https://github.com/Homebrew/homebrew-core). One PR per formula. And I will cloes this PR.

**Pull request closed without merging**